### ISSUE TITLE
custom training dataset, ChatDataset - understanding parameter train_on_input

### ISSUE THREAD

**Issue by oharapaGitHub** [pytorch/torchtune#1534]
I have followed the tutorial https://pytorch.org/torchtune/main/tutorials/chat.html, using it as reference and managed to fine tune both LLama2, and LLama3 with a custom dataset. (Lora single device)

The tutorial has proven very helpful but there is one part where I am confused.  In the tutorial (which I followed), a message_converter function is provided which takes the parameter train_on_input (below).
```
def message_converter(sample: Mapping[str, Any], train_on_input: bool) -> List[Message]:
    input_msg = sample["input"]
    output_msg = sample["output"]

    user_message = Message(
        role="user",
        content=input_msg,
        masked=not train_on_input,  # Mask if not training on prompt
    )
    assistant_message = Message(
        role="assistant",
        content=output_msg,
        masked=False,
    )
    # A single turn conversation
    messages = [user_message, assistant_message]

    return messages
```

The code from the tutorial which references the message_converter, custom_datset, instantiates and returns a ChatDataset object, where no value is passed for train_on_input.  Looking into the source for [ChatDataset](https://github.com/pytorch/torchtune/blob/main/torchtune/datasets/_chat.py#L73), the parameter train_on_input is false by default.

Based on this, and reviewing the message_converter, is the behaviour intended by the tutorial:

- when the role of message is user, set the 'masked' parameter to true (the negative of train_on_input, which is default false)
- when the role of message is assistant, always set the 'masked' parameter to false.

It would be great if this can be clarified, as I found it fairly confusing

**Comment by JinchuLi2002** [comment]
I was having the similar confusions as well,
It's mentioned in 
```
    - If ``train_on_input`` is True, the prompt is used during training and
    contributes to the loss.
    - If ``train_on_input`` is False, the prompt is masked out (tokens replaced with -100)
```

My understanding is the following:
1. Fine-tuning is always based on text-completion, though there are other fancy types like "chatDataset" etc, they are ultimately using text-completion with extra steps. e.g. concatenate your system & user & assistant messages and add some tags in between, etc.
2. Thus, for example if your training data looks like  "user: write xxx algorithm in Python. assistant: <some python code>". If you set `train_on_input` to True, and after training if your inference prompt looks like "user: write xxx algorithm in", the model will more likely continue by writing "Python" , rather than Java, c++, etc.

**Comment by oharapaGitHub** [comment]
@JinchuLi2002, thanks for the response, and sharing the understanding. It has helped make my understanding of train_on_input much clearer.  The same pattern looks to apply to Axolotl too [Axolotl Config Options](https://github.com/axolotl-ai-cloud/axolotl/blob/main/docs/config.qmd).

As an aside, you might be interested in the paper, [Instruction Tuning With Loss Over Instructions](https://arxiv.org/pdf/2405.14394), it highlights - use case dependent - it can be of benefit to have the loss applied to the instruction and output, over output alone.
